src/main/java/org/sonarqube/gradle/SonarQubePlugin.java

```diff
@@ -88,13 +92,36 @@ private void configureTask(SonarQubeTask sonarQubeTask, Project project, Map<Str
     conventionMapping.map("properties", () -> new SonarPropertyComputer(actionBroadcastMap, project)
       .computeSonarProperties());
 
-    Callable<Iterable<? extends Task>> testTasks = () -> project.getAllprojects().stream()
+    sonarQubeTask.mustRunAfter(getJavaTestTasks(project));
+    sonarQubeTask.dependsOn(getJavaCompileTasks(project));
+    sonarQubeTask.mustRunAfter(getJacocoTasks(project));
+    sonarQubeTask.dependsOn(getAndroidCompileTasks(project));
+  }
+
+  private static Callable<Iterable<? extends Task>> getJacocoTasks(Project project) {
+    return () -> project.getAllprojects().stream()
+      .filter(p -> p.getPlugins().hasPlugin(JacocoPlugin.class) && !p.getExtensions().getByType(SonarQubeExtension.class).isSkipProject())
+      .map(p -> p.getTasks().withType(JacocoReport.class))
+      .flatMap(Collection::stream)
+      .collect(Collectors.toList());
```

Comment:
This is not reliable.

1) Tasks might be added even after `sonarqube` plugin is applied. Then the newly added tasks won't be added to the dependency

2) There must not be dependencies among tasks between different projects. See https://docs.gradle.org/current/userguide/cross_project_publications.html#sec:simple-sharing-artifacts-between-projects

Reply:
The `Callable` is resolved only after the project is evaluated, when Gradle needs to figure out what tasks to run and in which order. We've used this for long time without any reported problems. 
About 2., I agree it's not ideal, but because of the way the SonarQube scanner works, it needs to run once for the root project and collect properties for all sub projects. So what we do is apply the plugin to the root project and depend on the subprojects.

Reply:
> after the project is evaluated

The projects might be evaluated in different order.

> it needs to run once for the root project and collect properties for all sub projects. So what we do is apply the plugin to the root project and depend on the subprojects

You might use configurations to share artifacts between projects or implement something like https://github.com/gradle/gradle/tree/master/subprojects/docs/src/samples/jvm-multi-project-with-code-coverage

Reply:
Thanks for the example. We can definitely improve things and I will keep it in mind but it's out of scope for these tickets. Here we are just changing the dependencies using the same system as before.

Reply:
Also, in my opinion things are not fundamentally different. The main difference is that we try to support multi-project out of the box.

### JaCoCo plugin
The JaCoCo plugin should be applied to every project. This will add report tasks and an extension to the project and to its `test` tasks. The report task sets a `mustRunAfter` the `test`.

The example with the multi-project then recommends that we:
* Create a custom task in the root project
* This task collects data from the JaCoCo extensions added to the `test` task in subprojects
* Add a dependency to the `test` tasks in the subprojects (for which a JaCoCo extension was added)

### Sonarqube plugin
The sonarqube plugin should be applied to a single project. This will add the task to the root project and an extension to all subprojects.

For multi-project, a `dependsOn` or `mustRunAfter` is added to tasks of interest in all subprojects. The sonarqube task will then collect properties from all subprojects by accessing its extension and the other tasks.